// @antv/x6/lib/registry/port-label-layout/side.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bottom = exports.top = exports.right = exports.left = exports.manual = void 0;
var util_1 = require("./util");
exports.manual = function (portPosition, elemBBox, args) { return util_1.toResult({ position: elemBBox.getTopLeft() }, args); };
exports.left = function (portPosition, elemBBox, args) {
    return util_1.toResult({
        position: { x: -15, y: 0 },
        attrs: { '.': { y: '.3em', 'text-anchor': 'end' } },
    }, args);
};
exports.right = function (portPosition, elemBBox, args) {
    return util_1.toResult({
        position: { x: 15, y: 0 },
        attrs: { '.': { y: '.3em', 'text-anchor': 'start' } },
    }, args);
};
exports.top = function (portPosition, elemBBox, args) {
    return util_1.toResult({
        position: { x: 0, y: -15 },
        attrs: { '.': { 'text-anchor': 'middle' } },
    }, args);
};
exports.bottom = function (portPosition, elemBBox, args) {
    return util_1.toResult({
        position: { x: 0, y: 15 },
        attrs: { '.': { y: '.6em', 'text-anchor': 'middle' } },
    }, args);
};
