// @antv/x6/lib/util/string/format.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.titleCase = exports.sentenceCase = exports.pathCase = exports.dotCase = exports.constantCase = exports.pascalCase = exports.capitalize = exports.upperFirst = exports.lowerFirst = exports.upperCase = exports.lowerCase = exports.snakeCase = exports.startCase = exports.kebabCase = exports.camelCase = void 0;
var lodash_es_1 = require("lodash-es");
Object.defineProperty(exports, "camelCase", { enumerable: true, get: function () { return lodash_es_1.camelCase; } });
Object.defineProperty(exports, "kebabCase", { enumerable: true, get: function () { return lodash_es_1.kebabCase; } });
Object.defineProperty(exports, "startCase", { enumerable: true, get: function () { return lodash_es_1.startCase; } });
Object.defineProperty(exports, "snakeCase", { enumerable: true, get: function () { return lodash_es_1.snakeCase; } });
Object.defineProperty(exports, "lowerCase", { enumerable: true, get: function () { return lodash_es_1.lowerCase; } });
Object.defineProperty(exports, "upperCase", { enumerable: true, get: function () { return lodash_es_1.upperCase; } });
Object.defineProperty(exports, "lowerFirst", { enumerable: true, get: function () { return lodash_es_1.lowerFirst; } });
Object.defineProperty(exports, "upperFirst", { enumerable: true, get: function () { return lodash_es_1.upperFirst; } });
Object.defineProperty(exports, "capitalize", { enumerable: true, get: function () { return lodash_es_1.capitalize; } });
var lodash_es_2 = require("lodash-es");
// @see: https://medium.com/@robertsavian/javascript-case-converters-using-lodash-4f2f964091cc
function pascalCase(str) {
    return lodash_es_2.startCase(lodash_es_2.camelCase(str)).replace(/ /g, '');
}
exports.pascalCase = pascalCase;
function constantCase(str) {
    return lodash_es_2.upperCase(str).replace(/ /g, '_');
}
exports.constantCase = constantCase;
function dotCase(str) {
    return lodash_es_2.lowerCase(str).replace(/ /g, '.');
}
exports.dotCase = dotCase;
function pathCase(str) {
    return lodash_es_2.lowerCase(str).replace(/ /g, '/');
}
exports.pathCase = pathCase;
function sentenceCase(str) {
    return lodash_es_2.upperFirst(lodash_es_2.lowerCase(str));
}
exports.sentenceCase = sentenceCase;
function titleCase(str) {
    return lodash_es_2.startCase(lodash_es_2.camelCase(str));
}
exports.titleCase = titleCase;
