// @antv/x6/lib/util/dom/path.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var elem_test_1 = require("./elem.test");
var vector_1 = require("./vector");
var vectorizer_1 = require("./vectorizer");
var path_1 = require("./path");
describe('Dom', function () {
    describe('path', function () {
        var _a = elem_test_1.setupTest(), svgContainer = _a.svgContainer, svgPath = _a.svgPath, svgGroup = _a.svgGroup, svgCircle = _a.svgCircle, svgEllipse = _a.svgEllipse, svgPolygon = _a.svgPolygon, svgText = _a.svgText, svgRectangle = _a.svgRectangle, svgGroup1 = _a.svgGroup1, svgGroup2 = _a.svgGroup2, svgGroup3 = _a.svgGroup3, svgPath2 = _a.svgPath2, svgPath3 = _a.svgPath3;
        afterAll(function () { return elem_test_1.clearnTest(); });
        describe('#toPath', function () {
            it('should convert SVGPathElement', function () {
                var path = vector_1.createVector('path', { d: 'M 100 50 L 200 150' });
                expect(path.convertToPath().getAttribute('d')).toBe('M 100 50 L 200 150');
            });
        });
        describe('#toPathData', function () {
            function roundPathData(pathData) {
                return pathData != null
                    ? pathData
                        .split(' ')
                        .map(function (command) {
                        var number = parseInt(command, 10);
                        if (isNaN(number))
                            return command;
                        return number.toFixed(0);
                    })
                        .join(' ')
                    : null;
            }
            it('should throw an exception on convert an invalid SvgElement', function () {
                expect(function () {
                    var group = vector_1.createVector('<group/>');
                    path_1.toPathData(group.node);
                }).toThrowError();
            });
            it('should convert SVGPathElement', function () {
                var path = vector_1.createVector('path', { d: 'M 100 50 L 200 150' });
                expect(path.convertToPathData()).toEqual('M 100 50 L 200 150');
            });
            it('should convert SVGLineElement', function () {
                var line = vector_1.createVector('line', {
                    x1: 100,
                    y1: 50,
                    x2: 200,
                    y2: 150,
                });
                expect(line.convertToPathData()).toEqual('M 100 50 L 200 150');
            });
            it('should convert SVGRectElement', function () {
                var rect = vector_1.createVector('rect', {
                    x: 100,
                    y: 50,
                    width: 200,
                    height: 150,
                });
                expect(rect.convertToPathData()).toEqual('M 100 50 H 300 V 200 H 100 V 50 Z');
            });
            it('should convert SVGRectElement with `rx` and `ry` attributes', function () {
                var rect = vector_1.createVector('<rect/>', {
                    x: 100,
                    y: 50,
                    width: 200,
                    height: 150,
                    rx: 200,
                    ry: 200,
                });
                expect(rect.convertToPathData()).toEqual('M 100 125 v 0 a 100 75 0 0 0 100 75 h 0 a 100 75 0 0 0 100 -75 v 0 a 100 75 0 0 0 -100 -75 h 0 a 100 75 0 0 0 -100 75 Z');
            });
            it('should convert SVGCircleElement', function () {
                var circle = vector_1.createVector('circle', { cx: 100, cy: 50, r: 50 });
                expect(roundPathData(circle.convertToPathData())).toEqual('M 100 0 C 127 0 150 22 150 50 C 150 77 127 100 100 100 C 72 100 50 77 50 50 C 50 22 72 0 100 0 Z');
            });
            it('should convert SVGEllipseElement', function () {
                var ellipse = vector_1.createVector('ellipse', {
                    cx: 100,
                    cy: 50,
                    rx: 100,
                    ry: 50,
                });
                expect(roundPathData(ellipse.convertToPathData())).toEqual('M 100 0 C 155 0 200 22 200 50 C 200 77 155 100 100 100 C 44 100 0 77 0 50 C 0 22 44 0 100 0 Z');
            });
            it('should convert SVGPolygonElement', function () {
                var polygon = vector_1.createVector('polygon', {
                    points: '200,10 250,190 160,210',
                });
                expect(polygon.convertToPathData()).toEqual('M 200 10 L250 190 L160 210 Z');
            });
            it('should convert SVGPolylineElement', function () {
                var polyline = vector_1.createVector('polyline', {
                    points: '100,10 200,10 150,110',
                });
                expect(polyline.convertToPathData()).toEqual('M 100 10 L200 10 L150 110');
            });
        });
        describe('#normalizePath', function () {
            it('should return this for any SVGElement', function () {
                expect(vector_1.createVector(svgPath).normalizePath()).toBeInstanceOf(vectorizer_1.Vectorizer);
                expect(vector_1.createVector(svgPath2).normalizePath()).toBeInstanceOf(vectorizer_1.Vectorizer);
                expect(vector_1.createVector(svgPath3).normalizePath()).toBeInstanceOf(vectorizer_1.Vectorizer);
                expect(vector_1.createVector(svgContainer).normalizePath()).toBeInstanceOf(vectorizer_1.Vectorizer);
                expect(vector_1.createVector(svgGroup).normalizePath()).toBeInstanceOf(vectorizer_1.Vectorizer);
                expect(vector_1.createVector(svgCircle).normalizePath()).toBeInstanceOf(vectorizer_1.Vectorizer);
                expect(vector_1.createVector(svgEllipse).normalizePath()).toBeInstanceOf(vectorizer_1.Vectorizer);
                expect(vector_1.createVector(svgPolygon).normalizePath()).toBeInstanceOf(vectorizer_1.Vectorizer);
                expect(vector_1.createVector(svgText).normalizePath()).toBeInstanceOf(vectorizer_1.Vectorizer);
                expect(vector_1.createVector(svgRectangle).normalizePath()).toBeInstanceOf(vectorizer_1.Vectorizer);
                expect(vector_1.createVector(svgGroup1).normalizePath()).toBeInstanceOf(vectorizer_1.Vectorizer);
                expect(vector_1.createVector(svgGroup2).normalizePath()).toBeInstanceOf(vectorizer_1.Vectorizer);
                expect(vector_1.createVector(svgGroup3).normalizePath()).toBeInstanceOf(vectorizer_1.Vectorizer);
            });
            it('shoule normalize path "d" attribute', function () {
                expect(vector_1.createVector(svgPath).normalizePath().node.hasAttribute('d')).toBe(true);
                expect(vector_1.createVector(svgPath2).normalizePath().node.hasAttribute('d')).toBe(true);
                expect(vector_1.createVector(svgPath3).normalizePath().node.hasAttribute('d')).toBe(true);
                expect(vector_1.createVector(svgPath).normalizePath().attr('d')).toEqual('M 10 10');
                expect(vector_1.createVector(svgPath2).normalizePath().attr('d')).toEqual('M 100 100 C 100 100 0 150 100 200 Z');
                expect(vector_1.createVector(svgPath3).normalizePath().attr('d')).toEqual('M 0 0');
            });
            it('should only normalize SVGPathElement', function () {
                expect(vector_1.createVector(svgContainer).normalizePath().node.hasAttribute('d')).toBe(false);
                expect(vector_1.createVector(svgGroup).normalizePath().node.hasAttribute('d')).toBe(false);
                expect(vector_1.createVector(svgCircle).normalizePath().node.hasAttribute('d')).toBe(false);
                expect(vector_1.createVector(svgEllipse).normalizePath().node.hasAttribute('d')).toBe(false);
                expect(vector_1.createVector(svgPolygon).normalizePath().node.hasAttribute('d')).toBe(false);
                expect(vector_1.createVector(svgText).normalizePath().node.hasAttribute('d')).toBe(false);
                expect(vector_1.createVector(svgRectangle).normalizePath().node.hasAttribute('d')).toBe(false);
                expect(vector_1.createVector(svgGroup1).normalizePath().node.hasAttribute('d')).toBe(false);
                expect(vector_1.createVector(svgGroup2).normalizePath().node.hasAttribute('d')).toBe(false);
                expect(vector_1.createVector(svgGroup3).normalizePath().node.hasAttribute('d')).toBe(false);
            });
        });
        describe('#createSlicePathData', function () {
            it('should return the path string of a part of sector', function () {
                expect(path_1.createSlicePathData(5, 10, 0, Math.PI / 2))
                    .toBe('M10,0A10,10 0 0,1 6.123233995736766e-16,10L3.061616997868383e-16,5A5,5 0 0,0 5,0Z');
            });
        });
    });
});
