// @antv/x6/lib/graph/scroller.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScrollerManager = void 0;
var types_1 = require("../types");
var base_1 = require("./base");
var ScrollerManager = /** @class */ (function (_super) {
    __extends(ScrollerManager, _super);
    function ScrollerManager() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(ScrollerManager.prototype, "widgetOptions", {
        get: function () {
            return this.options.scroller;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScrollerManager.prototype, "pannable", {
        get: function () {
            return this.widgetOptions && this.widgetOptions.pannable === true;
        },
        enumerable: false,
        configurable: true
    });
    ScrollerManager.prototype.init = function () {
        this.widget = this.graph.hook.createScroller();
        this.autoSetCursor();
        this.startListening();
        if (this.widget) {
            this.widget.center();
        }
    };
    ScrollerManager.prototype.startListening = function () {
        this.graph.on('blank:mousedown', this.preparePanning, this);
        this.graph.on('node:unhandled:mousedown', this.preparePanning, this);
        this.graph.on('edge:unhandled:mousedown', this.preparePanning, this);
    };
    ScrollerManager.prototype.stopListening = function () {
        this.graph.off('blank:mousedown', this.preparePanning, this);
        this.graph.off('node:unhandled:mousedown', this.preparePanning, this);
        this.graph.off('edge:unhandled:mousedown', this.preparePanning, this);
    };
    ScrollerManager.prototype.preparePanning = function (_a) {
        var e = _a.e;
        if (this.widget &&
            this.pannable &&
            types_1.ModifierKey.test(e, this.widgetOptions.modifiers) &&
            this.graph.hook.allowPanning(e)) {
            this.widget.startPanning(e);
        }
    };
    ScrollerManager.prototype.autoSetCursor = function () {
        var _a;
        var options = this.widgetOptions;
        if (options.cursor == null) {
            var cursor = options.pannable ? 'grab' : '';
            (_a = this.widget) === null || _a === void 0 ? void 0 : _a.setCursor(cursor, { silent: true });
        }
    };
    ScrollerManager.prototype.enablePanning = function () {
        if (!this.pannable) {
            this.widgetOptions.pannable = true;
            if (types_1.ModifierKey.equals(this.graph.options.scroller.modifiers, this.graph.options.selecting.modifiers)) {
                this.graph.selection.disableRubberband();
            }
        }
        this.autoSetCursor();
    };
    ScrollerManager.prototype.disablePanning = function () {
        if (this.pannable) {
            this.widgetOptions.pannable = false;
        }
        this.autoSetCursor();
    };
    ScrollerManager.prototype.lock = function () {
        if (this.widget) {
            this.widget.lock();
        }
    };
    ScrollerManager.prototype.unlock = function () {
        if (this.widget) {
            this.widget.unlock();
        }
    };
    ScrollerManager.prototype.update = function () {
        if (this.widget) {
            this.widget.update();
        }
    };
    ScrollerManager.prototype.setCursor = function (cursor) {
        if (this.widget) {
            this.widget.setCursor(cursor);
        }
    };
    ScrollerManager.prototype.resize = function (width, height) {
        if (this.widget) {
            this.widget.resize(width, height);
        }
    };
    ScrollerManager.prototype.dispose = function () {
        if (this.widget) {
            this.widget.dispose();
        }
        this.stopListening();
    };
    __decorate([
        base_1.Base.dispose()
    ], ScrollerManager.prototype, "dispose", null);
    return ScrollerManager;
}(base_1.Base));
exports.ScrollerManager = ScrollerManager;
