// @antv/x6/lib/shape/standard/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createShape = exports.labelAttr = exports.bodyAttr = exports.getMarkup = void 0;
var base_1 = require("../base");
var util_1 = require("../../util");
function getMarkup(tagName, selector) {
    if (selector === void 0) { selector = 'body'; }
    return [
        {
            tagName: tagName,
            selector: selector,
        },
        {
            tagName: 'text',
            selector: 'label',
        },
    ];
}
exports.getMarkup = getMarkup;
exports.bodyAttr = {
    fill: '#ffffff',
    stroke: '#333333',
    strokeWidth: 2,
};
exports.labelAttr = {
    fontSize: 14,
    fill: '#333333',
    refX: '50%',
    refY: '50%',
    textAnchor: 'middle',
    textVerticalAnchor: 'middle',
};
function createShape(shape, config, options) {
    var _a;
    if (options === void 0) { options = {}; }
    var defaults = {
        constructorName: shape,
        markup: getMarkup(shape, options.selector),
        attrs: (_a = {},
            _a[shape] = {
                fill: '#ffffff',
                stroke: '#333333',
                strokeWidth: 2,
            },
            _a),
    };
    var base = options.parent || base_1.Base;
    return base.define(util_1.ObjectExt.merge(defaults, config, { shape: shape }));
}
exports.createShape = createShape;
