// pncui-workflow/src/router/routers.js
import Vue from 'vue'
import Router from 'vue-router'
Vue.use(Router)

export const constantRouterMap = [
    {
        path: '/login',
        meta: {title: '登录', noCache: true},
        component: (resolve) => require([`eladmin-ui/src/views/login`], resolve),
        hidden: true
    },
    {
        path: '/sso/login',
        meta: {title: '统一登录', noCache: true},
        component: (resolve) => require(['eladmin-ui/src/views/ssologin'], resolve),
        hidden: true
    },
    {
        path: '/404',
        component: (resolve) => require(['eladmin-ui/src/views/features/404'], resolve),
        hidden: true
    },
    {
        path: '/401',
        component: (resolve) => require(['eladmin-ui/src/views/features/401'], resolve),
        hidden: true
    },
    {
        path: '/redirect',
        component: (resolve) => require(['eladmin-ui/src/layout'], resolve),
        hidden: true,
        children: [
            {
                path: '/redirect/:path*',
                component: (resolve) => require(['eladmin-ui/src/views/features/redirect'], resolve)
            }
        ]
    },
    {
        path: '/',
        component: (resolve) => require(['eladmin-ui/src/layout'], resolve),
        redirect: '/dashboard',
        children: [
            {
                path: 'dashboard',
                component: (resolve) => require([`eladmin-ui/src/views/home`], resolve),
                name: 'Dashboard',
                meta: {title: '首页', icon: 'index', affix: true, noCache: true}
            }
        ]
    },
    {
        path: '/user',
        component: (resolve) => require([`eladmin-ui/src/layout/`], resolve),
        hidden: true,
        redirect: 'noredirect',
        children: [
            {
                path: 'center',
                component: (resolve) => require(['eladmin-ui/src/views/system/user/center'], resolve),
                name: '个人中心',
                meta: {title: '个人中心'}
            }
        ]
    }
]

export default new Router({
    mode: 'history',
    scrollBehavior: () => ({y: 0}),
    routes: constantRouterMap
})
